upgrade/v122to130/upgrade.go

```diff
@@ -0,0 +1,89 @@
+package v122to130
+
+import (
+	"context"
+	"github.com/sirupsen/logrus"
+
+	"github.com/pkg/errors"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta1"
+	lhclientset "github.com/longhorn/longhorn-manager/k8s/pkg/client/clientset/versioned"
+)
+
+const (
+	upgradeLogPrefix = "upgrade from v1.2.2 to v1.3.0: "
+)
+
+func UpgradeCRs(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, upgradeLogPrefix+"UpgradeCRs failed")
+	}()
+	if err := upgradeEngines(namespace, lhClient); err != nil {
+		return err
+	}
+	return nil
+}
+
+func upgradeEngines(namespace string, lhClient *lhclientset.Clientset) (err error) {
+	defer func() {
+		err = errors.Wrapf(err, "upgrade engines failed")
+	}()
+
+	engines, err := lhClient.LonghornV1beta1().Engines(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	volumeEngineMap := map[string][]*v1beta1.Engine{}
+	for i := range engines.Items {
+		e := &engines.Items[i]
+		if e.Spec.VolumeName == "" {
+			// Cannot do anything in the upgrade path if this case happens.
```

Comment:
Do you mean the orphan engine? Could you add more explanation for `this case` in the comment? also add logging for this case.

Reply:
Yes. Once there is an orphan engine CR, we cannot do anything in the upgrade path. I will add this explanation in the comment.